near-sdk-macros/src/lib.rs

```diff
@@ -316,3 +371,81 @@ pub fn function_error(item: TokenStream) -> TokenStream {
         }
     })
 }
+
+/// The attribute macro `near_events` injects this macro and should be the used over this
+///
+/// This derive macro is used to inject the necessary wrapper and logic to auto format
+/// standard event logs. The other appropriate attribute macros are not injected with this macro.
+/// Required attributes below:
+/// ```ignore
+/// #[derive(near_sdk::serde::Serialize, std::clone::Clone)]
+/// #[serde(crate="near_sdk::serde")]
+/// #[serde(tag = "event", content = "data")]
+/// #[serde(rename_all="snake_case")]
+/// pub enum MyEvent {
+///     Event
+/// }
+/// ```
+#[proc_macro_derive(EventMetadata, attributes(event_meta))]
+pub fn derive_event_attributes(item: TokenStream) -> TokenStream {
+    if let Ok(input) = syn::parse::<ItemEnum>(item) {
+        let name = &input.ident;
+        // get standard and version from each attribute macro
+        let mut attr_error: u8 = 0;
+        let mut event_meta: Vec<proc_macro2::TokenStream> = vec![];
+        let _ = &input.variants.iter().for_each(|var| {
+            if let (Some(standard), Some(event)) = core_impl::get_event_args(var) {
+                let var_ident = &var.ident;
+                event_meta.push(quote! {
+                    #name::#var_ident { .. } => {(#standard.to_string(), #event.to_string())}
+                })
+            } else {
+                attr_error += 1;
+            }
+        });
+
+        // handle lifetimes, generics, and where clauses
+        let (impl_generics, type_generics, where_clause) = &input.generics.split_for_impl();
+
+        if attr_error > 0 {
+            return TokenStream::from(
+                syn::Error::new(
+                    Span::call_site(),
+                    "Near events must have `event_meta` attribute with `standard` and `version` fields for each event variant. Field values must be string literals",
+                )
+                .to_compile_error(),
+            );
+        }
+
+        TokenStream::from(quote! {
+
+            #[derive(near_sdk::serde::Serialize)]
+            #[serde(crate="near_sdk::serde")]
+            #[serde(rename_all="snake_case")]
+            struct EventBuilder #impl_generics #where_clause {
+                standard: String,
+                version: String,
+                #[serde(flatten)]
+                event_data: #name #type_generics
+            }
+            impl #impl_generics near_sdk::StandardEvent for #name #type_generics #where_clause {
+                fn format(&self) -> String {
+                    let (standard, version): (String, String) = match self {
+                        #(#event_meta),*
+                    };
+                    let event = EventBuilder {standard, version, event_data: self.clone() };
+                    near_sdk::serde_json::to_string(&event).unwrap_or_else(|_| near_sdk::env::abort())
+                }
+
+                fn emit(&self) {
+                    near_sdk::Event::emit(self.clone());
```

Comment:
Clone shouldn't be a required bound. You don't need an owned copy of the data to serialize. I haven't checked if this compiles, but it would look something like this:
```suggestion
            struct EventBuilder<'a> #impl_generics #where_clause {
                standard: String,
                version: String,
                #[serde(flatten)]
                event_data: &'a #name #type_generics
            }
            impl #impl_generics near_sdk::StandardEvent for #name #type_generics #where_clause {
                fn format(&self) -> String {
                    let (standard, version): (String, String) = match self {
                        #(#event_meta),*
                    };
                    let event = EventBuilder {standard, version, event_data: self };
                    near_sdk::serde_json::to_string(&event).unwrap_or_else(|_| near_sdk::env::abort())
                }

                fn emit(&self) {
                    near_sdk::Event::emit(self);
```